File: src/main/java/com/afollestad/polarupgradetool/XmlScanner.java
package com.afollestad.polarupgradetool;

/**
 * @author Aidan Follestad (afollestad)
 */
public class XmlScanner {

    private final int mInitialIndex;
    private int mIndex = 0;
    private StringBuilder mXml;
    private String mTagName;
    private String mTagValue;
    private int mValueStart;
    private int mValueEnd;
    private boolean mReachedEnd = false;

    public XmlScanner(StringBuilder xml) {
        mXml = xml;
        // Skip over the root tag
        mInitialIndex = mIndex = mXml.indexOf(">") + 1;
    }

//    public void updateXml(StringBuilder xml) {
//        mXml = xml;
//    }

    public String tagName() {
        return mTagName;
    }

    public String tagValue() {
        return mTagValue;
    }

    public boolean reachedEnd() {
        return mReachedEnd;
    }

    public String nextTag() {
        try {
            final int start = mXml.indexOf("<", mIndex);
            if (start < 0) {
                mReachedEnd = true;
                return null;
            } else if (mXml.charAt(start + 1) == '!') {
                // Skip comments
                mIndex = start + 1;
                return nextTag();
            }
            final int next = mXml.indexOf(">", start);
            final int firstSpace = mXml.indexOf(" ", start);
            if (firstSpace > next) {
                // Skip elements with no attributes
                mIndex = firstSpace + 1;
                return nextTag();
            }
            if (!mXml.substring(start, next).contains("name=")) {
                // Skip elements with no name attribute
                mIndex = next + 1;
                return nextTag();
            }
            mTagName = mXml.substring(start + 1, firstSpace);
            final String endFindStr = "</" + mTagName + ">";
            int end = mXml.indexOf(endFindStr, next + 1);
            if (end < 0) {
                // Didn't find an end to this tag, skip it
                mIndex = end + 1;
                return nextTag();
            }
            mValueStart = next + 1;
            mValueEnd = end;
            mTagValue = mXml.substring(mValueStart, mValueEnd);
            end += endFindStr.length();
            final String tag = mXml.substring(start, end);
            mIndex = end;
            return tag;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    public void reset() {
        mIndex = mInitialIndex;
        mReachedEnd = false;
    }

    public void setElementValue(String value) {
        mXml.replace(mValueStart, mValueEnd, value);
    }

    @Override
    public String toString() {
        return mXml.substring(mIndex);
    }
}

File: src/main/java/com/afollestad/polarupgradetool/xml/XmlScanner.kt
package com.afollestad.polarupgradetool.xml

/**
 * @author Aidan Follestad (afollestad)
 */
class XmlScanner(private val mXml: StringBuilder) {

    private var mIndex = 0
    private var mTagName: String = ""
    private var mTagValue: String = ""
    private var mTagStart: Int = 0
    private var mTagEnd: Int = 0
    private var mValueStart: Int = 0
    private var mValueEnd: Int = 0
    private var mReachedEnd = false

    fun tagName(): String {
        return mTagName
    }

    fun tagValue(): String {
        return mTagValue
    }

    fun reachedEnd(): Boolean {
        return mReachedEnd
    }

    fun currentTag(): String {
        return mXml.substring(mTagStart, mTagEnd)
    }

    fun nextTag(): String? {
        if (mReachedEnd) return null
        val next: Int
        val firstSpace: Int
        try {
            mTagStart = mXml.indexOf("<", mIndex)
            if (mTagStart < 0) {
                // No more tags in the file
                mReachedEnd = true
                return null
            } else if (mXml[mTagStart + 1] == '?') {
                // Skip header
                mIndex = mXml.indexOf("?>", mTagStart + 1) + 2
                return nextTag()
            } else if (mXml[mTagStart + 1] == '!') {
                // Skip comments
                mIndex = mXml.indexOf("-->", mTagStart + 1) + 3
                return nextTag()
            }

            next = mXml.indexOf(">", mTagStart)
            firstSpace = mXml.indexOf(" ", mTagStart)
            if (firstSpace == -1) {
                mReachedEnd = true
                return null
            } else if (firstSpace > next) {
                // Skip elements with no attributes
                mIndex = firstSpace + 1
                return nextTag()
            } else if (!mXml.substring(mTagStart, next).contains(" name=")) {
                // Skip elements with no name attribute
                mIndex = next + 1
                return nextTag()
            }

            mTagName = mXml.substring(mTagStart + 1, firstSpace)
            val endFindStr = "</$mTagName>"
            mTagEnd = mXml.indexOf(endFindStr, next + 1)
            if (mTagEnd < 0) {
                // Didn't find an end to this tag, skip it
                mIndex = mTagEnd + 1
                return nextTag()
            }
            mValueStart = next + 1
            mValueEnd = mTagEnd
            mTagValue = mXml.substring(mValueStart, mValueEnd)
            mTagEnd += endFindStr.length

            val tag = mXml.substring(mTagStart, mTagEnd)
            mIndex = mTagEnd
            return tag
        } catch (t: Throwable) {
            throw RuntimeException(t)
        }

    }

    //    public void reset() {
    //        mIndex = 0;
    //        mReachedEnd = false;
    //    }

    fun setElementValue(value: String) {
        val endFindStr = "</$mTagName>"
        mTagValue = value
        mXml.replace(mValueStart, mValueEnd, value)
        mValueEnd = mValueStart + mTagValue.length
        mTagEnd = mValueEnd + endFindStr.length
        mIndex = mTagEnd
    }

    override fun toString(): String {
        return mXml.substring(mIndex)
    }
}